docs/reference/cat/anomaly-detectors.asciidoc

```diff
@@ -0,0 +1,280 @@
+
+`data.out_of_order_timestamps`, `doot`, `dataOutOfOrderTimestamps`:::
+include::{docdir}/ml/ml-shared.asciidoc[tag=out-of-order-timestamp-count]
+
+`data.processed_fields`, `dpf`, `dataProcessedFields`:::
+include::{docdir}/ml/ml-shared.asciidoc[tag=processed-field-count]
+
+`data.processed_records`, `dpr`, `dataProcessedRecords`:::
+(Default)
+include::{docdir}/ml/ml-shared.asciidoc[tag=processed-record-count]
+
+`data.sparse_buckets`, `dsb`, `dataSparseBuckets`:::
+include::{docdir}/ml/ml-shared.asciidoc[tag=sparse-bucket-count]
+
+`forecasts.memory.avg`, `fmavg`, `forecastsMemoryAvg`:::
+The average memory usage in bytes for forecasts related to the {anomaly-job}.
+  
+`forecasts.memory.max`, `fmmax`, `forecastsMemoryMax`:::
+The maximum memory usage in bytes for forecasts related to the {anomaly-job}.
+
+`forecasts.memory.min`, `fmmin`, `forecastsMemoryMin`:::
+The minimum memory usage in bytes for forecasts related to the {anomaly-job}.
+
+`forecasts.memory.total`, `fmt`, `forecastsMemoryTotal`:::
+The total memory usage in bytes for forecasts related to the {anomaly-job}.                      
+  
+`forecasts.records.avg`, `fravg`, `forecastsRecordsAvg`:::
+The average number of `model_forecast` documents written for forecasts related
+to the {anomaly-job}.
+
+`forecasts.records.max`, `frmax`, `forecastsRecordsMax`:::
+The maximum number of `model_forecast` documents written for forecasts related
+to the {anomaly-job}.
+
+`forecasts.records.min`, `frmin`, `forecastsRecordsMin`:::
+The minimum number of `model_forecast` documents written for forecasts related
+to the {anomaly-job}.
+
+`forecasts.records.total`, `frt`, `forecastsRecordsTotal`:::
+The total number of `model_forecast` documents written for forecasts related to
+the {anomaly-job}.                         
+                                                   
+`forecasts.time.avg`, `ftavg`, `forecastsTimeAvg`:::
+The average runtime in milliseconds for forecasts related to the {anomaly-job}.
+
+`forecasts.time.max`, `ftmax`, `forecastsTimeMax`:::
+The maximum runtime in milliseconds for  forecasts related to the {anomaly-job}.
+
+`forecasts.time.min`, `ftmin`, `forecastsTimeMin`:::
+The minimum runtime in milliseconds for forecasts related to the {anomaly-job}.
+
+`forecasts.time.total`, `ftt`, `forecastsTimeTotal`:::
+The total runtime in milliseconds for forecasts related to the {anomaly-job}.
+
+`forecasts.total`, `ft`, `forecastsTotal`:::
+(Default)
+include::{docdir}/ml/ml-shared.asciidoc[tag=forecast-total]
+
+`id`:::
+(Default)
+include::{docdir}/ml/ml-shared.asciidoc[tag=job-id-anomaly-detection]
+
+`model.bucket_allocation_failures`, `mbaf`, `modelBucketAllocationFailures`:::
+include::{docdir}/ml/ml-shared.asciidoc[tag=bucket-allocation-failures-count]
+
+`model.by_fields`, `mbf`, `modelByFields`:::
+include::{docdir}/ml/ml-shared.asciidoc[tag=total-by-field-count]
+
+`model.bytes`, `mb`, `modelBytes`:::
```

Comment:
@benwtrent I've updated the API to use pluralized forecasts.* and buckets.* headings.
@droberts195 Should we pluralize the model-related headings too?  The current description for this stat seems to imply the answer will be yes:
```suggestion
`models.bytes`, `mb`, `modelsBytes`:::
```

Reply:
The stat this is reporting is `model_size_stats.model_bytes` from the job stats.  So it's singular in the field names.  In the [current docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.6/ml-get-job-stats.html) the description has "models" plural and "model" singular:

> (long) The number of bytes of memory used by the **models**. This is the maximum value since the last time the **model** was persisted. If the job is closed, this value indicates the latest size.

Whether model or models is correct is an interesting academic question.  Certainly at the lowest level of the C++ code there are multiple models, but we could argue these are components of a single top level model for the entire job.  We only report the stats for this top level model, so it's reasonable that we document it as a single model.  Even if the top level model consists of 1000 low level models there's no way for the end user to get any stats about these low level models through the Elasticsearch APIs so it doesn't really help to refer to them.

I think the pragmatic approach is to take singular or plural from the field names returned by the get job stats API.  These are also persisted in stored documents and it would be a major piece of work to change them now, which cannot be justified as it would mean we would spend months doing it instead of adding useful functionality.

So therefore I think we should go with `model` singular and `forecasts` plural.

`buckets`, `records` and `fields` in the `_cat` API have in many places been shortened from names containing `bucket_count`, `record_count` or `field_count` in the stats API, so there isn't the same degree of equivalence.

Reply:
Thanks for the explanation, @droberts195.  To be clear are you recommending that we revert to "bucket.*" instead of "buckets.*" columns in the cat datafeeds and cat anomaly detectors APIs?

Reply:
> To be clear are you recommending that we revert to "bucket." instead of "buckets." columns in the cat datafeeds and cat anomaly detectors APIs?

No, bucket/buckets is not directly comparable between the cat APIs and the stats APIs.  So I think it's OK to leave it as-is.